File: oneWordStory.py
def oneWordStory(words, sentences):
    ans = []
    picked = {}

    # a dictionary to keep track what indices in words were picked already
    for i in range(len(words)):
        picked[i] = False

    for i in range(len(sentences)):

        ans.append(sentences[i])
        lastWord = sentences[i][sentences[i].rfind(' ')+1:]
        # print lastWord
        maxDistance = -1
        selectedWord = ""
        index = None

        for j in range(len(words)):
            if picked[j] == True:
                continue
            if maxDistance == -1:
                maxDistance = levenshtein(lastWord, words[j])
                print maxDistance
                # print words[j]
                selectedWord = words[j]
                index = j
                continue

            maxL = levenshtein(lastWord, words[j])
            if maxL > maxDistance:
                maxDistance = maxL
                selectedWord = words[j]
                index = j
            elif maxDistance == maxL:
                if words[j] > selectedWord:
                    selectedWord = words[j]
                    index = j

        picked[index] = True
        ans.append(selectedWord)
    print ans
    return ' '.join(ans)


def levenshtein(s1, s2):
    if len(s1) < len(s2):
        return levenshtein(s2, s1)

    # len(s1) >= len(s2)
    if len(s2) == 0:
        return len(s1)

    previous_row = range(len(s2) + 1)
    for i, c1 in enumerate(s1):
        current_row = [i + 1]
        for j, c2 in enumerate(s2):
            # j+1 instead of j since previous_row and current_row are one character longer
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1       # than s2
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        previous_row = current_row

    return previous_row[-1]

File: oneWordStory.js
function oneWordStory(words, sentences) {
    let ans = [];

    // a map to keep track which indices in words are picked
    let picked = {};
    words.map((e, i) => picked[i] = false)

    for (let i = 0; i < sentences.length; i++) {

        ans.push(sentences[i]);
        let lastWord = sentences[i].substr(sentences[i].lastIndexOf(' ') + 1);
        let max;
        let selectedWord;
        let index;

        for (let j = 0; j < words.length; j++) {
            // word at index j was picked, ignore it
            if (picked[j]) continue;

            if (max == undefined) {
                max = levenshteinDistance(lastWord, words[j]);
                selectedWord = words[j];
                index = j;
                continue;
            }

            let maxDiff = levenshteinDistance(lastWord, words[j]);
            if (maxDiff > max) {
                max = maxDiff;
                selectedWord = words[j];
                index = j;

            } else if (maxDiff == max) {
                if (words[j] > selectedWord) {
                    selectedWord = words[j];
                    index = j;
                }
            }
        }

        // update index of word that was just picked
        picked[index] = true;
        // console.log(`picked = ${JSON.stringify(picked)}`)
        ans.push(selectedWord);
    }

    return ans.join(' ')

}

// get Levenshtein distance btw two strings
function levenshteinDistance(str1, str2) {
    str1 = str1.split(``);
    str2 = str2.split(``);

    let distance = [],
        l = str1.length,
        m = str2.length,
        i, j;

    for (i = 0; i <= l; i++) distance[i] = [i]
    for (j = 0; j <= m; j++) distance[0][j] = j

    for (i = 1; i <= l; i++) {
        for (j = 1; j <= m; j++) {
            distance[i][j] = Math.min(distance[i - 1][j] + 1, distance[i][j - 1] + 1, distance[i - 1][j - 1] +
                ((str1[i - 1] == str2[j - 1]) ? 0 : 1));
        }
    }
    return distance[l][m];

};